# Credit Risk Evaluator

In [1]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import precision_score, recall_score, f1_score

## Retrieve the Data

The data is located in the Challenge Files Folder:

* `lending_data.csv`

Import the data using Pandas. Display the resulting dataframe to confirm the import was successful.

In [2]:
# Import the data
file = 'Resources/lending_data.csv'
df = pd.read_csv(file)

In [3]:
df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77536 entries, 0 to 77535
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   loan_size         77536 non-null  float64
 1   interest_rate     77536 non-null  float64
 2   borrower_income   77536 non-null  int64  
 3   debt_to_income    77536 non-null  float64
 4   num_of_accounts   77536 non-null  int64  
 5   derogatory_marks  77536 non-null  int64  
 6   total_debt        77536 non-null  int64  
 7   loan_status       77536 non-null  int64  
dtypes: float64(3), int64(5)
memory usage: 4.7 MB


## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*Replace the text in this markdown cell with your predictions, and be sure to provide justification for your guess.*

## Split the Data into Training and Testing Sets

In [5]:
# Split the data into X_train, X_test, y_train, y_test
X = df.drop('loan_status', axis=1)
y = df['loan_status']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Create, Fit and Compare Models

Create a Logistic Regression model, fit it to the data, and print the model's score. Do the same for a Random Forest Classifier. You may choose any starting hyperparameters you like. 

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the designated markdown cell.

### Logistic Regression model

In [6]:
# Train a Logistic Regression model and print the model score

# Create a Logistic Regression model
lrm = LogisticRegression(random_state=42)

# Fit the Logistic Regression model to the data
lrm.fit(X_train, y_train)

# Make predictions on the test data
y_pred = lrm.predict(X_test)

# Calculate precision, recall, and F1 score
lrm_precision = precision_score(y_test, y_pred)
lrm_recall = recall_score(y_test, y_pred)
lrm_f1 = f1_score(y_test, y_pred)

# Print the scores
print(f"Logistic Regression score: {lrm.score(X_test, y_test)}")
print(f"Logistic Regression precision: {lrm_precision}")
print(f"Logistic Regression recall: {lrm_recall}")
print(f"Logistic Regression F1 score: {lrm_f1}")

Logistic Regression score: 0.9928424039205571
Logistic Regression precision: 0.8754863813229572
Logistic Regression recall: 0.9054325955734407
Logistic Regression F1 score: 0.8902077151335311


### Random Forest Classifier

In [7]:
# Train a Random Forest Classifier model and print the model score

# Create a Random Forest Classifier
rfc = RandomForestClassifier(n_estimators=100, random_state=42)

# Fit the Random Forest Classifier to the data
rfc.fit(X_train, y_train)

# Make predictions on the test data
y_pred = rfc.predict(X_test)

# Calculate precision, recall, and F1 score
rfc_precision = precision_score(y_test, y_pred)
rfc_recall = recall_score(y_test, y_pred)
rfc_f1 = f1_score(y_test, y_pred)

# Print the scores
print(f"Random Forest Classifier score: {rfc.score(X_test, y_test)}")
print(f"Random Forest Classifier precision: {rfc_precision}")
print(f"Random Forest Classifier recall: {rfc_recall}")
print(f"Random Forest Classifier F1 score: {rfc_f1}")

Random Forest Classifier score: 0.9925844725303069
Random Forest Classifier precision: 0.8715953307392996
Random Forest Classifier recall: 0.9014084507042254
Random Forest Classifier F1 score: 0.8862512363996043


# Which model performed better?

Which model performed better? How does that compare to your prediction? Replace the text in this markdown cell with your answers to these questions.

In [8]:
# Create a dictionary with the evaluation metrics
data = {
    "Model": ["Logistic Regression", "Random Forest Classifier"],
    "Score": [lrm.score(X_test, y_test), rfc.score(X_test, y_test)],
    "Precision": [lrm_precision, rfc_precision],
    "Recall": [lrm_recall, rfc_recall],
    "F1 Score": [lrm_f1, rfc_f1]
}

# Create a pandas dataframe with the evaluation metrics
df = pd.DataFrame(data)

# Set the "Model" column as the index
df.set_index("Model", inplace=True)

# Display the dataframe
print(df)


                             Score  Precision    Recall  F1 Score
Model                                                            
Logistic Regression       0.992842   0.875486  0.905433  0.890208
Random Forest Classifier  0.992584   0.871595  0.901408  0.886251


### I predicted that the Random Forest Classifier would perform better due to its robustness. However, based on the evaluation metrics, the Logistic Regression model outperformed the Random Forest Classifier. The Logistic Regression model achieved a higher score of 0.9925 compared to the Random Forest Classifier's score of 0.9885, and a slightly higher F1 score of 0.8902 compared to 0.886251

In [9]:
"""
                             Score  Precision    Recall  F1 Score
Model                                                            
Logistic Regression       0.992842   0.875486  0.905433  0.890208
Random Forest Classifier  0.992584   0.871595  0.901408  0.886251
"""

'\n                             Score  Precision    Recall  F1 Score\nModel                                                            \nLogistic Regression       0.992842   0.875486  0.905433  0.890208\nRandom Forest Classifier  0.992584   0.871595  0.901408  0.886251\n'